src/Core/Aetherling/Passes/ThroughputModifications.hs

```diff
@@ -1,339 +1,508 @@
+-- However, behavior is undefined if requesting an invalid throughput multiplier
+-- that doesn't match the divisibility requirements such as those for linebuffer.
+-- Not going to speed these up by wrapping them in a map because that would
+-- cause inappropriate behavior in the linebuffer. The linebuffer has complex
+-- internal state and mapping over it will cause it to behave unintuitively.
+-- For example, an LineBuffer [2] [3] [300] T_Int will reading in 2 int per clock
+-- and emit two overlapping windows of consecutive ints.
+-- A Map 2 (LineBuffer [1] [3] [300] T_Int) will emit two windows, but the
+-- ints in the windows will not be consecutive. The first window will have ints
+-- that came in 0, 2, and 4 (0-indexed) and the second window will have ints
+-- indexed 1, 3, and 5.
+
+-- speed up inner most dimension first. Once you have an entire row input per
+-- clock, then speed up the outer dimensions like number of rows per clock.
+-- see increaseLBPxPerClock for more information. 
+-- NOTE: increaseLBPxPerClock expects dimensions from innermost to outermost
+-- and LineBuffer stores them in the opposite order. Thus, they must be
+-- reversed when calling the helper function.
+attemptSpeedUp requestedMult (LineBuffer p w img t bc) =
   (LineBuffer (reverse reversedNewP) w img t bc, actualMult)
   where
     (reversedNewP, actualMult) =
-      increaseLBPxPerClock (reverse p) (reverse img) throughMult
--- don't increase the arr size, duplicate it so it can be fed to other maps
-attemptSpeedUp throughMult op@(Constant_Int _) = (MapOp throughMult op, throughMult)
-attemptSpeedUp throughMult op@(Constant_Bit _) = (MapOp throughMult op, throughMult)
--- not going to change SLen in as, if 1, don't want to cause it to become
--- fractional
-attemptSpeedUp throughMult (SequenceArrayRepack (sLenIn, oldArrLenIn)
+      increaseLBPxPerClock (reverse p) (reverse img) requestedMult
+
+-- could decrease sLenIn and sLenOut as increase throughput, but not going
+-- to do that as don't want to deal with fractional sequence lengths,
+-- which could happen if dividing sLenIn or sLenOut 
+attemptSpeedUp requestedMult (SequenceArrayRepack (sLenIn, oldArrLenIn)
                               (sLenOut, oldArrLenOut) t) =
-  (SequenceArrayRepack (sLenIn, oldArrLenIn * throughMult)
-    (sLenOut, oldArrLenOut * throughMult) t, throughMult)
-attemptSpeedUp throughMult op@(ArrayReshape _ _) =
-  (MapOp throughMult op, throughMult)
-attemptSpeedUp throughMult op@(DuplicateOutputs _ _) =
-  (MapOp throughMult op, throughMult)
-
--- is this an ok inverse of our decision to only underutil for slowdown of seq operators?
--- should this try to speed up children before itself?
-attemptSpeedUp throughMult (MapOp par innerOp) | isComb innerOp =
-  (MapOp (par*throughMult) innerOp, throughMult)
-attemptSpeedUp throughMult (MapOp par innerOp) =
-  let (spedUpInnerOp, actualMult) = attemptSpeedUp throughMult innerOp
```

Comment:
This seems to be the case where the innerOp __IS NOT__ parallelizable via a map.  But the guard checks to see if innerOp is combinational.  An innerOp can definitely be non-combinational, but also parallelizable.  In fact, innerOp is already being used in a map, SO IT MUST BE PARALLELIZABLE.  As a result, I think we can drop the second case of speeding up a map.  I think only the first one is necessary.  

In other words, what is a situation where map k1 innerOp is valid, but map k1*m innerOp would not be?  I can't think of one.